### PR TITLE
openjdk: Ensure rpath for libraries includes server/lib

### LINKED ISSUE #111068: openjdk: dlopen/rpath issues when loading an internal openjdk .so after loading an external .so
## `brew config`

```
$ brew config
HOMEBREW_VERSION: 3.6.1-72-gdf9f878
ORIGIN: https://github.com/Homebrew/brew
HEAD: df9f8786cac3c4755c186ac5bbe63e2198862c20
Last commit: 2 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: 934055f356a95bb0b9221a9f5a8bb45dc8e476a3
Core tap last commit: 44 minutes ago
Core tap branch: master
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_EDITOR: vi
HOMEBREW_MAKE_JOBS: 10
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: 10-core 64-bit arm_firestorm_icestorm
Clang: 14.0.0 build 1400
Git: 2.37.0 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.6-arm64
CLT: 14.0.0.0.1.1661618636
Xcode: N/A
Rosetta 2: false
$
```

## `brew doctor`
```
$ brew doctor
Your system is ready to brew.
$  
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

Use the [pyroscope java agent](https://github.com/pyroscope-io/pyroscope-java) with [opennms](https://github.com/OpenNMS/opennms) for continuous profiling to debug application issues. This loads the [async-profiler](https://github.com/jvm-profiling-tools/async-profiler/) and its .so to get profiling information out of the JVM. I have the same problem when I try to load `async-profiler` directly. I'm doing this on an M1 Mac using Homebrew OpenJDK.

I have a simple reproducer, `ManagementTest` and all of the details I'll share will be from there.

If I load either the pyroscope agent or the async-profiler into Homebrew-built OpenJDK 11.0.16.1 or 18.0.2.1, I get this failure:
```
Exception in thread "main" java.lang.UnsatisfiedLinkError: no management in java.library.path: [/Users/dgregor/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2673)
	at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:830)
	at java.base/java.lang.System.loadLibrary(System.java:1873)
	at java.management/java.lang.management.ManagementFactory.lambda$static$8(ManagementFactory.java:1020)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.management/java.lang.management.ManagementFactory.<clinit>(ManagementFactory.java:1019)
	at ManagementTest.main(ManagementTest.java:6)
```

### What happened (include all command output)?

(note that I had to hit ctrl-C to terminate the JVM when loading the pyroscope agent)
```
$ /opt/homebrew/opt/openjdk@11/bin/java -agentpath:/Users/dgregor/Downloads/async-profiler-2.8.3-macos/build/libasyncProfiler.so=start,event=cpu,file=profile.html -cp . ManagementTest
Profiling started
Exception in thread "main" java.lang.UnsatisfiedLinkError: no management in java.library.path: [/Users/dgregor/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2673)
	at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:830)
	at java.base/java.lang.System.loadLibrary(System.java:1873)
	at java.management/java.lang.management.ManagementFactory.lambda$static$8(ManagementFactory.java:1020)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.management/java.lang.management.ManagementFactory.<clinit>(ManagementFactory.java:1019)
	at ManagementTest.main(ManagementTest.java:6)
$ /opt/homebrew/opt/openjdk@18/bin/java -agentpath:/Users/dgregor/Downloads/async-profiler-2.8.3-macos/build/libasyncProfiler.so=start,event=cpu,file=profile.html -cp . ManagementTest
Profiling started
Exception in thread "main" java.lang.UnsatisfiedLinkError: no management in system library path: /opt/homebrew/Cellar/openjdk/18.0.2.1/libexec/openjdk.jdk/Contents/Home/lib
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2408)
	at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:848)
	at java.base/java.lang.System.loadLibrary(System.java:2015)
	at java.management/java.lang.management.ManagementFactory.lambda$loadNativeLib$8(ManagementFactory.java:1025)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at java.management/java.lang.management.ManagementFactory.loadNativeLib(ManagementFactory.java:1024)
	at java.management/java.lang.management.ManagementFactory.<clinit>(ManagementFactory.java:1019)
	at ManagementTest.main(ManagementTest.java:6)
$ /opt/homebrew/opt/openjdk@11/bin/java -javaagent:/Users/dgregor/Downloads/pyroscope.jar -cp . ManagementTest
2022-09-18 16:42:07.026 [INFO] We recommend specifying application name via env variable PYROSCOPE_APPLICATION_NAME
2022-09-18 16:42:07.034 [INFO] For now we chose the name for you and it's javaspy.VeRo3kppTE6W-aS19qBBug
2022-09-18 16:42:07.034 [WARN] PYROSCOPE_SERVER_ADDRESS is not defined, using http://localhost:4040
2022-09-18 16:42:07.707 [INFO] Profiling started
Exception in thread "main" java.lang.UnsatisfiedLinkError: no management in java.library.path: [/Users/dgregor/Library/Java/Extensions, /Library/Java/Extensions, /Network/Library/Java/Extensions, /System/Library/Java/Extensions, /usr/lib/java, .]
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2673)
	at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:830)
	at java.base/java.lang.System.loadLibrary(System.java:1873)
	at java.management/java.lang.management.ManagementFactory.lambda$static$8(ManagementFactory.java:1020)
	at java.base/java.security.AccessController.doPrivileged(Native Method)
	at java.management/java.lang.management.ManagementFactory.<clinit>(ManagementFactory.java:1019)
	at ManagementTest.main(ManagementTest.java:6)
^CProfiling stopped after 0 seconds. No dump options specified
$ /opt/homebrew/opt/openjdk@18/bin/java -javaagent:/Users/dgregor/Downloads/pyroscope.jar -cp . ManagementTest
2022-09-18 16:42:13.893 [INFO] We recommend specifying application name via env variable PYROSCOPE_APPLICATION_NAME
2022-09-18 16:42:13.911 [INFO] For now we chose the name for you and it's javaspy.GYdlDc5eQD-4rP8oy7ZY8g
2022-09-18 16:42:13.911 [WARN] PYROSCOPE_SERVER_ADDRESS is not defined, using http://localhost:4040
2022-09-18 16:42:14.227 [INFO] Profiling started
Exception in thread "main" java.lang.UnsatisfiedLinkError: no management in system library path: /opt/homebrew/Cellar/openjdk/18.0.2.1/libexec/openjdk.jdk/Contents/Home/lib
	at java.base/java.lang.ClassLoader.loadLibrary(ClassLoader.java:2408)
	at java.base/java.lang.Runtime.loadLibrary0(Runtime.java:848)
	at java.base/java.lang.System.loadLibrary(System.java:2015)
	at java.management/java.lang.management.ManagementFactory.lambda$loadNativeLib$8(ManagementFactory.java:1025)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:318)
	at java.management/java.lang.management.ManagementFactory.loadNativeLib(ManagementFactory.java:1024)
	at java.management/java.lang.management.ManagementFactory.<clinit>(ManagementFactory.java:1019)
	at ManagementTest.main(ManagementTest.java:6)
^CProfiling stopped after 4 seconds. No dump options specified
$
```

If you run with `DYLD_PRINT_APIS=1` (see [this page for details](https://lowlevelbits.org/debugging-dyld/)), you get to see the detailed error is that `libmanagement.dylib` fails to load `libjvm.dylib`: `Library not loaded: '@rpath/libjvm.dylib'`. Note that `libjvm.dylib` is in `lib/server`, *not* directly in `lib`:
```
1969:dyld[1473]:       dlopen(libmanagement.dylib) => NULL, 'dlopen(/opt/homebrew/Cellar/openjdk@11/11.0.16.1/libexec/openjdk.jdk/Contents/Home/lib/libmanagement.dylib, 0x0001): Library not loaded: '@rpath/libjvm.dylib'
```

Details:
```
$ DYLD_PRINT_APIS=1 /opt/homebrew/opt/openjdk@11/bin/java -agentpath:/Users/dgregor/Downloads/async-profiler-2.8.3-macos/build/libasyncProfiler.so=start,event=cpu,file=profile.html -cp . ManagementTest 2>&1 | egrep -n 'dlopen.*(libasyncProfiler|libmanagement)'
143:dyld[1473]: dlopen("/Users/dgregor/Downloads/async-profiler-2.8.3-macos/build/libasyncProfiler.so", 0x00000001)
146:dyld[1473]:       dlopen(libasyncProfiler.so) => 0x2089f7360
150:dyld[1473]: dlopen("/Users/dgregor/Downloads/async-profiler-2.8.3-macos/build/libasyncProfiler.so", 0x00000081)
151:dyld[1473]:       dlopen(libasyncProfiler.so) => 0x2089f7360
1439:dyld[1473]: dlopen("/Users/dgregor/Downloads/async-profiler-2.8.3-macos/build/libasyncProfiler.so", 0x00000011)
1440:dyld[1473]:       dlopen(libasyncProfiler.so) => 0x2089f7360
1968:dyld[1473]: dlopen("/opt/homebrew/Cellar/openjdk@11/11.0.16.1/libexec/openjdk.jdk/Contents/Home/lib/libmanagement.dylib", 0x00000001)
1969:dyld[1473]:       dlopen(libmanagement.dylib) => NULL, 'dlopen(/opt/homebrew/Cellar/openjdk@11/11.0.16.1/libexec/openjdk.jdk/Contents/Home/lib/libmanagement.dylib, 0x0001): Library not loaded: '@rpath/libjvm.dylib'
1972:dyld[1473]: dlerror() => 'dlopen(/opt/homebrew/Cellar/openjdk@11/11.0.16.1/libexec/openjdk.jdk/Contents/Home/lib/libmanagement.dylib, 0x0001): Library not loaded: '@rpath/libjvm.dylib'
1977:dyld[1473]: dlopen("/opt/homebrew/Cellar/openjdk@11/11.0.16.1/libexec/openjdk.jdk/Contents/Home/lib/libmanagement.jnilib", 0x00000001)
1978:dyld[1473]:       dlopen(libmanagement.jnilib) => NULL, 'dlopen(/opt/homebrew/Cellar/openjdk@11/11.0.16.1/libexec/openjdk.jdk/Contents/Home/lib/libmanagement.jnilib, 0x0001): tried: '/opt/homebrew/Cellar/openjdk@11/11.0.16.1/libexec/openjdk.jdk/Contents/Home/lib/libmanagement.jnilib' (no such file)'
1979:dyld[1473]: dlerror() => 'dlopen(/opt/homebrew/Cellar/openjdk@11/11.0.16.1/libexec/openjdk.jdk/Contents/Home/lib/libmanagement.jnilib, 0x0001): tried: '/opt/homebrew/Cellar/openjdk@11/11.0.16.1/libexec/openjdk.jdk/Contents/Home/lib/libmanagement.jnilib' (no such file)'
$
```

### What did you expect to happen?

The application starts up properly with the agent setup.

It does if I use a patched copy of async-profiler with the patch in this issue: https://github.com/jvm-profiling-tools/async-profiler/issues/647 (see more information the reproduction section).

See also: https://github.com/Homebrew/homebrew-core/issues/66953

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
I'm going to do this with `async-profiler` because it is a bit simpler than Pyroscope (Pyroscope includes a patched version of `async-profiler`.

1. Download [async-profiler-2.8.3-macos.zip](https://github.com/jvm-profiling-tools/async-profiler/releases/download/v2.8.3/async-profiler-2.8.3-macos.zip) and extract it.
2. Create ManagementTest.java with the contents below.
3. Compile: `javac ManagementTest.java`
4. Run (adjust paths to JDK and `libasyncProfiler.so` as appropriate): `/opt/homebrew/opt/openjdk@11/bin/java -agentpath:/Users/dgregor/Downloads/async-profiler-2.8.3-macos/build/libasyncProfiler.so=start,event=cpu,file=profile.html -cp . ManagementTest`
5. Run the same thing, but with the environment variable DYLD_PRINT_APIS=1 set (see the "What happened" section above).


## ManagementTest.java

import java.lang.management.ManagementFactory;
import java.lang.management.RuntimeMXBean;

public class ManagementTest {
    public static void main(String[] argv) {
        final RuntimeMXBean runtime = ManagementFactory.getRuntimeMXBean();
        System.out.println("it worked");
    }
}
```

If I apply [this workaround](https://github.com/jvm-profiling-tools/async-profiler/commit/b5634b9d88304e64c0f77cf961119f240a979c38) from [#647: Patch rpath for Homebrew JDK](https://github.com/jvm-profiling-tools/async-profiler/issues/647) to `async-profiler`, things work fine. Pyroscope also works fine if I apply the same patch to their [patched version of async-profiler](https://github.com/pyroscope-io/async-profiler) and rebuild their `.jar` with updated versions of the `.so` and its `.sha1` file:

(note: this example was hacked up from my notes/history, so it might not be 100% correct)
```
git clone https://github.com/jvm-profiling-tools/async-profiler.git
cd async-profiler
cd git show b5634b9d88304e64c0f77cf961119f240a979c38 > /tmp/rpath.patch
cd ..
git clone https://github.com/pyroscope-io/async-profiler.git async-profiler-pyroscope
cd async-profiler-pyroscope
git patch < /tmp/rpath.patch
make FAT_BINARY=true
cp build/libasyncProfiler.so libasyncProfiler-macos.so
sha1sum libasyncProfiler-macos.so > libasyncProfiler-macos.so.sha1
cp /Users/dgregor/Downloads/pyroscope.jar pyroscope-hacked.jar
zip -n '.so' -u pyroscope-hacked.jar libasyncProfiler-macos.so libasyncProfiler-macos.so.sha1
```

I'm unfortunately not familiar enough with the details of Homebrew builds and the JDK build infrastructure to easily dig into what could be tweaked in https://github.com/iMichka/homebrew-core/blob/master/Formula/openjdk@11.rb to fix this.
```

### PR DESCRIPTION
**Pull request by deejgregor**
Also:
openjdk@11: Ensure rpath for libraries includes server/lib
openjdk@17: Ensure rpath for libraries includes server/lib

I couldn't build openjdk@8 because I have too new of a version of Xcode installed, so I did not include a change for that formula.

Closes: #111068

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

### CODE REVIEW

**Comment by deejgregor** [comment]
Question: I'm not sure if this needs a new revision as mentioned here: https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md#to-contribute-a-fix-to-the-foo-formula

**Comment by carlocab** [comment]
Yes, let's bump the `revision`.

You can also remove the `ignore_missing` call inside the `on_linux` blocks. Those will be fixed by the `-rpath` configuration.

File: Formula/openjdk.rb [hunk 1]
```diff
@@ -89,14 +88,15 @@ def install
...
-        --with-extra-ldflags=-headerpad_max_install_names
+        --with-extra-ldflags=-headerpad_max_install_names\ -Wl,-rpath,#{loader_path}/server
         --with-sysroot=#{MacOS.sdk_path}
       ]
     else
       %W[
         --with-x=#{HOMEBREW_PREFIX}
         --with-cups=#{HOMEBREW_PREFIX}
         --with-fontconfig=#{HOMEBREW_PREFIX}
+        --with-extra-ldflags=-Wl,-rpath,#{loader_path}/server"
```
Comment (Bo98) [inline Formula/openjdk.rb:99]: I'd personally have a ldflags array that we `join(" ")` so we don't have this duplication.

Comment (carlocab) [reply Formula/openjdk.rb]: Yea, I considered that too. Let's do that.

**Comment by carlocab** [comment]
Linkage still broken on Linux. The build may not be respecting `--with-extra-ldflags`.

**Comment by carlocab** [comment]
Here's the `RPATH` in `libmanagement.so` in the Linux bottle for `openjdk`:
```
$ORIGIN
/homebrew/linuxbrew/.linuxbrew/Cellar/openjdk/18.0.2.1_1/lib
/homebrew/linuxbrew/.linuxbrew/opt/gcc/lib/gcc/current
/homebrew/linuxbrew/.linuxbrew/opt/alsa-lib/lib
/homebrew/linuxbrew/.linuxbrew/opt/openssl@1.1/lib
/homebrew/linuxbrew/.linuxbrew/opt/bison/lib
/homebrew/linuxbrew/.linuxbrew/opt/krb5/lib
/homebrew/linuxbrew/.linuxbrew/opt/zlib/lib
/homebrew/linuxbrew/.linuxbrew/opt/gmp/lib
/homebrew/linuxbrew/.linuxbrew/opt/bdw-gc/lib
/homebrew/linuxbrew/.linuxbrew/opt/libtool/lib
/homebrew/linuxbrew/.linuxbrew/opt/libunistring/lib
/homebrew/linuxbrew/.linuxbrew/opt/ncurses/lib
/homebrew/linuxbrew/.linuxbrew/opt/readline/lib
/homebrew/linuxbrew/.linuxbrew/opt/libffi/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxcrypt/lib
/homebrew/linuxbrew/.linuxbrew/opt/guile/lib
/homebrew/linuxbrew/.linuxbrew/opt/icu4c/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxml2/lib
/homebrew/linuxbrew/.linuxbrew/opt/gettext/lib
/homebrew/linuxbrew/.linuxbrew/opt/libidn2/lib
/homebrew/linuxbrew/.linuxbrew/opt/libtasn1/lib
/homebrew/linuxbrew/.linuxbrew/opt/nettle/lib
/homebrew/linuxbrew/.linuxbrew/opt/p11-kit/lib
/homebrew/linuxbrew/.linuxbrew/opt/libevent/lib
/homebrew/linuxbrew/.linuxbrew/opt/libnghttp2/lib
/homebrew/linuxbrew/.linuxbrew/opt/expat/lib
/homebrew/linuxbrew/.linuxbrew/opt/unbound/lib
/homebrew/linuxbrew/.linuxbrew/opt/gnutls/lib
/homebrew/linuxbrew/.linuxbrew/opt/cups/lib
/homebrew/linuxbrew/.linuxbrew/opt/libpng/lib
/homebrew/linuxbrew/.linuxbrew/opt/bzip2/lib
/homebrew/linuxbrew/.linuxbrew/opt/freetype/lib
/homebrew/linuxbrew/.linuxbrew/opt/util-linux/lib
/homebrew/linuxbrew/.linuxbrew/opt/fontconfig/lib
/homebrew/linuxbrew/.linuxbrew/opt/isl/lib
/homebrew/linuxbrew/.linuxbrew/opt/mpfr/lib
/homebrew/linuxbrew/.linuxbrew/opt/libmpc/lib
/homebrew/linuxbrew/.linuxbrew/opt/lz4/lib
/homebrew/linuxbrew/.linuxbrew/opt/xz/lib
/homebrew/linuxbrew/.linuxbrew/opt/zstd/lib
/homebrew/linuxbrew/.linuxbrew/opt/binutils/lib
/homebrew/linuxbrew/.linuxbrew/opt/gcc/lib
/homebrew/linuxbrew/.linuxbrew/opt/libpthread-stubs/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxau/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxdmcp/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxcb/lib
/homebrew/linuxbrew/.linuxbrew/opt/libx11/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxext/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxrender/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxrandr/lib
/homebrew/linuxbrew/.linuxbrew/opt/libice/lib
/homebrew/linuxbrew/.linuxbrew/opt/libsm/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxt/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxfixes/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxi/lib
/homebrew/linuxbrew/.linuxbrew/opt/libxtst/lib
/homebrew/linuxbrew/.linuxbrew/lib
```
This is missing `$ORIGIN/server`, which means our flag isn't being respected.

**Comment by carlocab** [comment]
Ok, not sure what's going on with Linux, so I'll keep the `ignore_missing_libraries` for now. CC @Homebrew/linux in case anyone has any ideas.

**Comment by carlocab** [comment]
Dependent tests succeeded, but `openjdk@11` failed to build on Linux. To save CI time let's re-run but skip dependents.

**Review by carlocab — APPROVED** [review]
Thanks!

**Comment by danielnachun** [comment]
The `ignore_missing_libraries` is needed because OpenJDK finds these libraries using `dlopen` at runtime.  So to the best of my knowledge this is expected and not a problem.

**Comment by carlocab** [comment]
`ignore_missing_libraries` is something we carry around for essentially just OpenJDK, though, and it wouldn't be needed if we just had `$ORIGIN/server` in `RPATH`. It would be good to be able to get rid of it.

**Pull request closed without merging**